compiler/src/main/scala/org/scalajs/nscplugin/GenJSExports.scala

```diff
@@ -777,27 +777,32 @@ trait GenJSExports extends SubComponent { self: GenJSCode =>
 
     /** Generate the final forwarding call to the exported method. */
     private def genResult(exported: Exported, args: List[js.Tree],
-        static: Boolean)(implicit pos: Position) = {
+        static: Boolean)(implicit pos: Position): js.Tree = {
       val sym = exported.sym
-      val thisType =
-        if (sym.owner == ObjectClass) jstpe.ClassType(ir.Definitions.ObjectClass)
-        else encodeClassType(sym.owner)
-      val receiver =
-        if (static) genLoadModule(sym.owner)
-        else js.This()(thisType)
-      val call = {
-        if (isNonNativeJSClass(currentClassSym)) {
-          assert(sym.owner == currentClassSym.get, sym.fullName)
-          genApplyJSClassMethod(receiver, sym, args)
-        } else {
-          if (sym.isClassConstructor)
-            genApplyMethodStatically(receiver, sym, args)
-          else
-            genApplyMethod(receiver, sym, args)
-        }
+
+      def receiver = {
+        if (static)
+          genLoadModule(sym.owner)
+        else if (sym.owner == ObjectClass)
+          js.This()(jstpe.ClassType(ir.Definitions.ObjectClass))
+        else
+          js.This()(encodeClassType(sym.owner))
+      }
+
+      def boxIfNeeded(call: js.Tree): js.Tree = {
+        ensureBoxed(call,
+            enteringPhase(currentRun.posterasurePhase)(sym.tpe.resultType))
+      }
+
+      if (isNonNativeJSClass(currentClassSym)) {
+        assert(sym.owner == currentClassSym.get, sym.fullName)
+        boxIfNeeded(genApplyJSClassMethod(receiver, sym, args))
+      } else {
+        if (sym.isClassConstructor)
+          genNew(currentClassSym, sym, args)
```

Comment:
Why don't we need a box here? Because classes that are boxed never have exported constructors?

Reply:
No, it's because `genNew` already bypasses the logic that would have removed a box (in other words it already contains the box). In practice this only makes a difference if one exports the constructor of an `AnyVal`. Exporting the constructors of hijacked classes is not valid, because they have no constructors to begin with.

But the real reason to specialize the code path is that we generally avoid relying on the `resultType` of scalac class constructors, because they're the class in scalac but they must be `NoType` in our IR. And we definitely must not box to `undefined` here ;)